File: server/agent/src/llm/kaggle-fallback.js
const fetch = require('node-fetch');
require('dotenv').config();

const TIMEOUT_MS = 15000;

function getEndpoint() {
    const url = process.env.KAGGLE_ENDPOINT_URL;
    if (!url) return null;
    return url.replace(/\/$/, '');
}

async function generateResponse(systemPrompt, userMessage, history = []) {
    const endpoint = getEndpoint();
    if (!endpoint) {
        return { success: false, error: 'KAGGLE_ENDPOINT_URL not configured', source: 'kaggle' };
    }

    try {
        const controller = new AbortController();
        const timeout = setTimeout(() => controller.abort(), TIMEOUT_MS);

        const res = await fetch(`${endpoint}/generate`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                system_prompt: systemPrompt,
                message: userMessage,
                history: history
            }),
            signal: controller.signal
        });

        clearTimeout(timeout);

        if (!res.ok) {
            return { success: false, error: `Kaggle endpoint returned ${res.status}`, source: 'kaggle' };
        }

        const data = await res.json();
        return { success: true, text: data.text || data.response, source: 'kaggle' };
    } catch (error) {
        return { success: false, error: error.message, source: 'kaggle' };
    }
}

async function generateJSON(systemPrompt, userMessage) {
    const result = await generateResponse(
        systemPrompt + '\n\nYou MUST respond with valid JSON only. No markdown, no explanation.',
        userMessage
    );

    if (!result.success) return { ...result, source: 'kaggle' };

    try {
        
        let jsonText = result.text;
        const jsonMatch = jsonText.match(/```(?:json)?\s*([\s\S]*?)```/);
        if (jsonMatch) jsonText = jsonMatch[1];

        const parsed = JSON.parse(jsonText.trim());
        return { success: true, data: parsed, raw: jsonText, source: 'kaggle' };
    } catch (e) {
        return { success: false, error: `JSON parse failed: ${e.message}`, raw: result.text, source: 'kaggle' };
    }
}

async function healthCheck() {
    const endpoint = getEndpoint();
    if (!endpoint) return false;

    try {
        const controller = new AbortController();
        const timeout = setTimeout(() => controller.abort(), 5000);
        const res = await fetch(`${endpoint}/health`, { signal: controller.signal });
        clearTimeout(timeout);
        return res.ok;
    } catch {
        return false;
    }
}

module.exports = { generateResponse, generateJSON, healthCheck };

File: server_py/agent/llm/kaggle_fallback.py
"""
Kaggle Endpoint Fallback — Calls a remote Kaggle-hosted model.
"""

import json
import re
import httpx
from config import get_settings

TIMEOUT_S = 15  # 15 seconds


def _get_endpoint() -> str | None:
    settings = get_settings()
    url = settings.KAGGLE_ENDPOINT_URL
    if not url:
        return None
    return url.rstrip("/")


async def generate_response(
    system_prompt: str,
    user_message: str,
    history: list[dict] = None,
) -> dict:
    """Generate a text response via the Kaggle endpoint."""
    endpoint = _get_endpoint()
    if not endpoint:
        return {"success": False, "error": "KAGGLE_ENDPOINT_URL not configured", "source": "kaggle"}

    try:
        async with httpx.AsyncClient(timeout=TIMEOUT_S) as client:
            resp = await client.post(
                f"{endpoint}/generate",
                json={
                    "system_prompt": system_prompt,
                    "message": user_message,
                    "history": history or [],
                },
            )

        if resp.status_code != 200:
            return {"success": False, "error": f"Kaggle endpoint returned {resp.status_code}", "source": "kaggle"}

        data = resp.json()
        return {"success": True, "text": data.get("text") or data.get("response", ""), "source": "kaggle"}

    except Exception as e:
        return {"success": False, "error": str(e), "source": "kaggle"}


async def generate_json(system_prompt: str, user_message: str) -> dict:
    """Generate a JSON response via Kaggle."""
    result = await generate_response(
        system_prompt + "\n\nYou MUST respond with valid JSON only. No markdown, no explanation.",
        user_message,
    )

    if not result.get("success"):
        return {**result, "source": "kaggle"}

    try:
        json_text = result["text"]
        match = re.search(r"```(?:json)?\s*([\s\S]*?)```", json_text)
        if match:
            json_text = match.group(1)

        parsed = json.loads(json_text.strip())
        return {"success": True, "data": parsed, "raw": json_text, "source": "kaggle"}

    except Exception as e:
        return {"success": False, "error": f"JSON parse failed: {e}", "raw": result.get("text"), "source": "kaggle"}


async def health_check() -> bool:
    """Check if the Kaggle endpoint is reachable."""
    endpoint = _get_endpoint()
    if not endpoint:
        return False

    try:
        async with httpx.AsyncClient(timeout=5) as client:
            resp = await client.get(f"{endpoint}/health")
            return resp.status_code == 200
    except Exception:
        return False
